File: src/designpatterns/facade/FacadeCPP/converter_facade.cpp
#include "audio_mixer.hpp"
#include "bitrate_reader.hpp"
#include "codec_factory.hpp"

class ConverterFacade {
public:
    string convertVideo(const string& filename, const string& format) {
        cout << "[Facade] Starting conversion..." << endl;

        VideoFile file(filename);
        auto srcCodec = CodecFactory::extract(file);
        unique_ptr<Codec> dstCodec;

        if (format == "mp4") {
            dstCodec = make_unique<MPEG4Compression>();
        } else {
            dstCodec = make_unique<OGGCompression>();
        }

        string buffer = BitrateReader::read(filename, *srcCodec);
        string result = BitrateReader::convert(buffer, *dstCodec);

        AudioMixer mixer;
        result = mixer.fix(result);

        cout << "[Facade] Completed conversion..." << endl;
        return result;
    }
};

File: src/designpatterns/facade/FacadeTS/converter_facade.ts
import { BitrateReader } from './bitrate_reader.ts';
import { CodecFactory } from './codec_factory.ts';
import { VideoFile } from './video_file.ts';
import {
    type Codec, MPEG4Compression, OGGCompression 
} from './codec.ts';
import { AudioMixer } from './audio_mixer.ts';

class ConverterFacade {
    convertVideo(filename: string, format: string): string {
        console.log('[Facade] Starting conversion...');

        const file = new VideoFile(filename);
        const srcCodec = CodecFactory.extract(file);
        const dstCodec = (format == 'mp4') ?
            new MPEG4Compression() : new OGGCompression();

        const buffer = BitrateReader.read(filename, srcCodec);
        const result = BitrateReader.convert(buffer, dstCodec);

        const mixer = new AudioMixer();
        const finalResult = mixer.fix(result);

        console.log('[Facade] Completed conversion...');
        return finalResult;
    }

    // Client code
    public static main(): void {
        console.log('[Client] I can easily convert file')
        const converter = new ConverterFacade();
        const mp4 = converter.convertVideo('funny_cat.ogg', 'mp4');
        console.log(`[Client] Now I got a ${mp4}.mp4`);
    }
}

ConverterFacade.main();